File: src/main/java/cz/wa/secretagent/view/renderer/model/AnimatedModelDrawer.java
package cz.wa.secretagent.view.renderer.model;

import java.util.List;

import org.apache.commons.math3.geometry.euclidean.twod.Vector2D;
import org.springframework.beans.factory.annotation.Required;

import cz.wa.secretagent.view.texture.DrawBounds;
import cz.wa.secretagent.view.texture.TextureToDraw;
import cz.wa.secretagent.world.entity.Entity;
import cz.wa.secretagent.world.entity.HasDuration;
import cz.wa.secretagent.world.entity.HasModelAngle;
import cz.wa.secretagent.world.entity.HasTime;
import cz.wa.secretagent.worldinfo.WorldHolder;
import secretAgent.view.model.AnimatedModel;
import secretAgent.view.renderer.model.AbstractModelDrawer;

/**
 * Draws animated model. 
 *  
 * @author Ondrej Milenovsky
 */
public class AnimatedModelDrawer extends AbstractModelDrawer<AnimatedModel> {
    private static final long serialVersionUID = 6233405165118049099L;

    private WorldHolder worldHolder;

    @Override
    public void draw(AnimatedModel model, Entity entity, Vector2D pos, double scale) {
        // time from the entity or from level
        long timeMs;
        if (entity instanceof HasTime) {
            timeMs = ((HasTime) entity).getTimeMs();
        } else {
            timeMs = worldHolder.getWorld().getLevelMap().getTimeMs();
        }

        // duration from the entity or from the model
        long durationMs;
        if (entity instanceof HasDuration) {
            durationMs = ((HasDuration) entity).getDurationMs();
        } else {
            durationMs = model.getDurationMs();
        }

        // draw the texture
        List<TextureToDraw> textures = model.getTextures();
        TextureToDraw tex = getFrame(timeMs, durationMs, textures);
        if (entity instanceof HasModelAngle) {
            primitivesDrawer.drawTexture(tex, pos, new DrawBounds(model.getBounds()), scale,
                    ((HasModelAngle) entity).getModelAngle());
        } else {
            primitivesDrawer.drawTexture(tex, pos, new DrawBounds(model.getBounds()), scale);
        }
    }

    public WorldHolder getWorldHolder() {
        return worldHolder;
    }

    @Required
    public void setWorldHolder(WorldHolder worldHolder) {
        this.worldHolder = worldHolder;
    }

}

File: src/main/kotlin/secretAgent/view/renderer/model/AnimatedModelDrawer.kt
package secretAgent.view.renderer.model

import cz.wa.secretagent.worldinfo.WorldHolder
import org.apache.commons.math3.geometry.euclidean.twod.Vector2D
import secretAgent.view.model.AnimatedModel
import secretAgent.view.renderer.DrawBounds
import secretAgent.world.entity.Entity
import secretAgent.world.entity.HasDuration
import secretAgent.world.entity.HasModelAngle
import secretAgent.world.entity.HasTime

/**
 * Draws animated model.
 *
 * @author Ondrej Milenovsky
 */
class AnimatedModelDrawer : AbstractModelDrawer<AnimatedModel>() {

    lateinit var worldHolder: WorldHolder

    override fun draw(model: AnimatedModel, entity: Entity?, pos: Vector2D, scale: Double) {
        // time from the entity or from level
        val timeMs = when (entity) {
            is HasTime -> (entity as HasTime).timeMs
            else -> worldHolder.world.levelMap.timeMs
        }

        // duration from the entity or from the model
        val durationMs = when (entity) {
            is HasDuration -> (entity as HasDuration).durationMs
            else -> model.durationMs
        }

        // draw the texture
        val textures = model.textures
        val tex = AbstractModelDrawer.getFrame(timeMs, durationMs, textures!!)!!
        if (entity is HasModelAngle)
            primitivesDrawer.drawTexture(tex, pos, DrawBounds(model.bounds), scale, (entity as HasModelAngle).modelAngle)
        else
            primitivesDrawer.drawTexture(tex, pos, DrawBounds(model.bounds), scale)
    }

    companion object {
        private const val serialVersionUID = 6233405165118049099L
    }
}